soccer/planning/InterpolatedPath.cpp

```diff
@@ -190,69 +195,76 @@ void InterpolatedPath::draw(DebugDrawer* const debug_drawer,
-        return false;
-    }
-
-    targetVelOut = direction * linearSpeed;
-    */
-    if (waypoints.size() == 0 || waypoints.size() == 1) {
+    if (waypoints.size() < 2) {
         return std::nullopt;
     }
-    if (t < waypoints[0].time) {
-        debugThrow(
-            invalid_argument("The start time should not be less than zero"));
+    if (t <= waypoints.front().time) {
+        return waypoints.front().instant();
+    } else if (t > waypoints.back().time) {
+        return std::nullopt;
+    } else if (t == waypoints.back().time) {
+        return waypoints.back().instant();
     }
 
-    int i = 0;
-    while (waypoints[i].time <= t) {
-        if (waypoints[i].time == t) {
-            return RobotInstant(waypoints[i].instant);
-        }
-        i++;
-        if (i == size()) {
-            return std::nullopt;
+    std::vector<Entry>::const_iterator prev_it = waypoints.begin();
+    std::vector<Entry>::const_iterator next_it = waypoints.begin();
+    while (next_it != waypoints.end()) {
+        prev_it = next_it;
+        next_it++;
+
+        if (next_it->time >= t) {
+            break;
         }
     }
-    RJ::Seconds deltaT = (waypoints[i].time - waypoints[i - 1].time);
-    if (deltaT == RJ::Seconds::zero()) {
-        return RobotInstant(waypoints[i].instant);
-    }
-    float constant = (t - waypoints[i - 1].time) / deltaT;
 
-    return RobotInstant(MotionInstant(waypoints[i - 1].pos() * (1 - constant) +
-                                          waypoints[i].pos() * (constant),
-                                      waypoints[i - 1].vel() * (1 - constant) +
-                                          waypoints[i].vel() * (constant)));
+    assert(prev_it != waypoints.end());
+    assert(next_it != waypoints.end());
+
+    Entry prev_entry = *prev_it;
+    Entry next_entry = *next_it;
+
+    RJ::Seconds dt = next_entry.time - prev_entry.time;
+    if (dt == RJ::Seconds(0)) {
+        return next_entry.instant();
+    }
+    RJ::Seconds elapsed = t - prev_entry.time;
+
+    // s in [0, 1] is the interpolation factor.
+    double s = elapsed / dt;
+
+    Pose pose_0 = prev_entry.pose;
+    Pose pose_1 = next_entry.pose;
+    Twist tangent_0 = prev_entry.vel * RJ::numSeconds(dt);
+    Twist tangent_1 = next_entry.vel * RJ::numSeconds(dt);
+
+    // Cubic interpolation
+    Pose interpolated_pose =
+        Pose(Eigen::Vector3d(pose_0) * (2 * s * s * s - 3 * s * s + 1) +
+             Eigen::Vector3d(tangent_0) * (s * s * s - 2 * s * s + s) +
+             Eigen::Vector3d(pose_1) * (-2 * s * s * s + 3 * s * s) +
+             Eigen::Vector3d(tangent_1) * (s * s * s - s * s));
+
+    Twist interpolated_twist =
+        Twist(Eigen::Vector3d(pose_0) * (6 * s * s - 6 * s) +
+              Eigen::Vector3d(tangent_0) * (3 * s * s - 4 * s + 1) +
+              Eigen::Vector3d(pose_1) * (-6 * s * s + 6 * s) +
+              Eigen::Vector3d(tangent_1) * (3 * s * s - 2 * s)) /
+        RJ::numSeconds(dt);
```

Comment:
If there is a source on these equations, even if it's a wikipedia link, that's always good to throw into a comment.

Reply:
ack